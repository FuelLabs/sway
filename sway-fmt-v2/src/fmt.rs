use crate::utils::{
    indent_style::Shape, newline_style::apply_newline_style, program_type::insert_program_type,
};
use std::{path::Path, sync::Arc};
use sway_core::BuildConfig;

pub use crate::{
    config::manifest::Config,
    error::{ConfigError, FormatterError},
};

#[derive(Debug, Default)]
pub struct Formatter {
    pub shape: Shape,
    pub config: Config,
}

pub type FormattedCode = String;

pub trait Format {
    fn format(
        &self,
        formatted_code: &mut FormattedCode,
        formatter: &mut Formatter,
    ) -> Result<(), FormatterError>;
}

impl Formatter {
    pub fn from_dir(dir: &Path) -> Result<Self, ConfigError> {
        let config = match Config::from_dir(dir) {
            Ok(config) => config,
            Err(ConfigError::NotFound) => Config::default(),
            Err(e) => return Err(e),
        };
        let shape = Shape::default();
        Ok(Self { config, shape })
    }
    pub fn format(
        &mut self,
        src: Arc<str>,
        build_config: Option<&BuildConfig>,
    ) -> Result<FormattedCode, FormatterError> {
        let path = build_config.map(|build_config| build_config.canonical_root_module());
        let src_len = src.len();
        let module = sway_parse::parse_file(src, path)?;
        // Get parsed items
        let items = module.items;
        // Get the program type (script, predicate, contract or library)
        let program_type = module.kind;

        // Formatted code will be pushed here with raw newline stlye.
        // Which means newlines are not converted into system-specific versions until `apply_newline_style()`.
        // Use the length of src as a hint of the memory size needed for `raw_formatted_code`,
        // which will reduce the number of reallocations
        let mut raw_formatted_code = String::with_capacity(src_len);

        // Insert program type to the formatted code.
        insert_program_type(&mut raw_formatted_code, program_type)?;

        // Insert parsed & formatted items into the formatted code.
        let mut iter = items.iter().peekable();
        while let Some(item) = iter.next() {
            // format Annotated<ItemKind>
            item.format(&mut raw_formatted_code, self)?;
            if iter.peek().is_some() {
                raw_formatted_code.push('\n');
            }
        }

        let mut formatted_code = String::from(&raw_formatted_code);
        // Replace newlines with specified `NewlineStyle`
        apply_newline_style(
            self.config.whitespace.newline_style,
            &mut formatted_code,
            &raw_formatted_code,
        )?;
        if formatted_code.ends_with('\n') {
            formatted_code.pop();
        }

        Ok(formatted_code)
    }
}

#[cfg(test)]
mod tests {
    use super::Formatter;
    use crate::config::user_def::FieldAlignment;
    use std::sync::Arc;

    #[test]
    fn test_const() {
        let sway_code_to_format = r#"contract;
pub const TEST:u16=10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_struct_multiline_line_alignment() {
        let sway_code_to_format = r#"contract;
pub struct Foo<T, P> {
   barbazfoo: u64,
   baz  : bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo<T, P> {
    barbazfoo : u64,
    baz       : bool,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(40);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_single_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_single_line() {
        let sway_code_to_format = r#"contract;
pub enum Foo {
    bar: u64,
    baz: bool,
}
"#;
        let correct_sway_code = r#"contract;

pub enum Foo { bar: u64, baz: bool }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_struct_multi_line() {
        let sway_code_to_format = r#"contract;
pub struct Foo {
    bar: u64,
    baz: bool
}
"#;
        let correct_sway_code = r#"contract;

pub struct Foo {
    bar: u64,
    baz: bool,
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_enum_without_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: () }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue: (),
    Green: (),
    Red: (),
    Silver: (),
    Grey: (),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_enum_with_variant_alignment() {
        let sway_code_to_format = r#"contract;

enum Color {
    Blue: (), Green: (),
            Red: (),
    Silver: (),
                    Grey: (), }
        "#;
        let correct_sway_code = r#"contract;

enum Color {
    Blue   : (),
    Green  : (),
    Red    : (),
    Silver : (),
    Grey   : (),
}"#;

        // Creating a config with enum_variant_align_threshold that exceeds longest variant length
        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(20);

        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_abi_with_generics_and_attributes() {
        let sway_code_to_format = r#"contract;

abi StorageMapExample {
    #[storage(write)]fn insert_into_map1(key: u64, value: u64);

fn hello(key: u64, value: u64);
}"#;
        let correct_sway_code = r#"contract;

abi StorageMapExample {
    #[storage(write)]
    fn insert_into_map1(key: u64, value: u64);

    fn hello(key: u64, value: u64);
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }

    #[test]
    fn test_multi_items() {
        let sway_code_to_format = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;
        let correct_sway_code = r#"contract;

pub const TEST: u16 = 10;
pub const TEST1: u16 = 10;"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_ty_formatting() {
        let sway_code_to_format = r#"contract;

enum TestTy {
    Infer:
    _,
    Array : [u8;
    40],
    String:         str[
    4
    ],
    PathType     : root::
example::
    type,
    TupleNil: (),
    Tuple: (   u64, 
        u32
    ),
}"#;
        let correct_sway_code = r#"contract;

enum TestTy {
    Infer: _,
    Array: [u8; 40],
    String: str[4],
    PathType: root::example::type,
    TupleNil: (),
    Tuple: (u64, u32),
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code);
    }
    fn test_storage_without_alignment() {
        let sway_code_to_format = r#"contract;

storage{foo:Test=Test{},bar
: 
    Test=Test{}
, baz: u64 } 
"#;
        let correct_sway_code = r#"contract;

storage {
    foo: Test,
    bar: Test,
    baz: u64,
}"#;

        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_with_alignment() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage {
    long_var_name : Type1,
    var2          : Type2,
}"#;

        let mut formatter = Formatter::default();
        formatter.config.structures.field_alignment = FieldAlignment::AlignFields(50);
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_storage_single_line() {
        let sway_code_to_format = r#"contract;

storage {
 long_var_name: Type1=Type1{},
      var2: Type2=Type2{},
}
"#;
        let correct_sway_code = r#"contract;

storage { long_var_name: Type1, var2: Type2 }"#;
        let mut formatter = Formatter::default();
        formatter.config.structures.small_structures_single_line = true;
        formatter.config.whitespace.max_width = 300;
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_item_fn() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String {let greeting = 42;greeting.to_string()}
fn goodbye() -> usize {let farewell: usize = 5; farewell }"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String {
    let greeting = 42;
    greeting.to_string()
}

fn goodbye() -> usize {
    let farewell: usize = 5;
    farewell
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_same_line_where() {
        let sway_code_to_format = r#"contract;

pub fn hello( person: String ) -> String where T: Eq,{let greeting = 42;greeting.to_string()}"#;
        let correct_sway_code = r#"contract;

pub fn hello(person: String) -> String
where
    T: Eq,
{
    let greeting = 42;
    greeting.to_string()
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
    #[test]
    fn test_trait_and_super_trait() {
        let sway_code_to_format = r#"library traits;

trait Person{ fn name( self )->String;fn age( self )->usize; }
trait Student:Person {fn university(self) -> String;}
trait Programmer {fn fav_language(self) -> String;}
trait CompSciStudent: Programmer+Student {fn git_username(self) -> String;}"#;
        let correct_sway_code = r#"library traits;

trait Person {
    fn name(self) -> String;

    fn age(self) -> usize;
}

trait Student: Person {
    fn university(self) -> String;
}

trait Programmer {
    fn fav_language(self) -> String;
}

trait CompSciStudent: Programmer + Student {
    fn git_username(self) -> String;
}"#;
        let mut formatter = Formatter::default();
        let formatted_sway_code =
            Formatter::format(&mut formatter, Arc::from(sway_code_to_format), None).unwrap();
        assert_eq!(correct_sway_code, formatted_sway_code)
    }
}
