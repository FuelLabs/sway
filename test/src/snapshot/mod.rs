use anyhow::Result;
use libtest_mimic::{Arguments, Trial};
use normalize_path::NormalizePath;
use regex::Regex;
use std::{
    path::{Path, PathBuf},
    str::FromStr,
    sync::Once,
};

static FORC_COMPILATION: Once = Once::new();
static FORC_DOC_COMPILATION: Once = Once::new();

fn compile_forc() {
    let args = vec!["b", "--release", "-p", "forc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

fn compile_forc_doc() {
    let args = vec!["b", "--release", "-p", "forc-doc"];
    let o = std::process::Command::new("cargo")
        .args(args)
        .output()
        .unwrap();
    assert!(o.status.success());
}

pub(super) async fn run(filter_regex: Option<&regex::Regex>) -> Result<()> {
    let repo_root = PathBuf::from_str(&std::env::var("CARGO_MANIFEST_DIR")?)?
        .parent()
        .unwrap()
        .to_path_buf();
    let test_root = repo_root.join("test");
    let test_programs_dir = test_root.join("src/e2e_vm_tests/test_programs/");

    let args = Arguments {
        filter: filter_regex.as_ref().map(|filter| filter.to_string()),
        nocapture: true,
        ..Default::default()
    };

    let tests = discover_tests(&test_root)
        .into_iter()
        .map(|dir| {
            let name = dir
                .strip_prefix(&test_programs_dir)
                .unwrap()
                .display()
                .to_string();

            let repo_root = repo_root.clone();
            Trial::test(name, move || {
                let snapshot_toml =
                    std::fs::read_to_string(format!("{}/snapshot.toml", dir.display()))?;
                let snapshot_toml = toml::from_str::<toml::Value>(&snapshot_toml)?;
                let cmds = if let Some(cmds) = snapshot_toml.get("cmds") {
                    cmds.as_array()
                        .unwrap()
                        .iter()
                        .map(|v| v.as_str().unwrap())
                        .collect::<Vec<_>>()
                } else {
                    vec!["forc build --path {root}"]
                };

                let root = dir.strip_prefix(&repo_root).unwrap().display().to_string();

                use std::fmt::Write;
                let mut snapshot = String::new();

                for cmd in cmds {
                    let cmd = cmd.replace("{root}", &root);

                    let _ = writeln!(&mut snapshot, "> {}", cmd);

                    let mut last_output: Option<String> = None;

                    for cmd in cmd.split("|") {
                        let cmd = cmd.trim();

                        // known commands
                        let cmd = if let Some(cmd) = cmd.strip_prefix("forc doc ") {
                            FORC_DOC_COMPILATION.call_once(|| {
                                compile_forc_doc();
                            });
                            format!("target/release/forc-doc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("forc ") {
                            FORC_COMPILATION.call_once(|| {
                                compile_forc();
                            });
                            format!("target/release/forc {cmd} 1>&2")
                        } else if let Some(cmd) = cmd.strip_prefix("grep ") {
                            let arg = cmd.trim();
                            if let Some(l) = last_output.take() {
                                let mut new_output = String::new();
                                for line in l.lines() {
                                    if line.contains(arg) {
                                        new_output.push_str(line);
                                        new_output.push('\n');
                                    }
                                }
                                last_output = Some(new_output);
                            }
                            continue;
                        } else {
                            panic!("Not supported. Possible commands: forc")
                        };

                        let o = duct::cmd!("bash", "-c", cmd.clone())
                            .dir(repo_root.clone())
                            .stderr_to_stdout()
                            .stdout_capture();

                        let o = if let Some(last_output) = last_output.as_ref() {
                            o.stdin_bytes(last_output.as_bytes())
                        } else {
                            o
                        };

                        let o = o.env("COLUMNS", "10").unchecked().start().unwrap();
                        let o = o.wait().unwrap();
                        last_output = Some(clean_output(&format!(
                            "exit status: {}\noutput:\n{}",
                            o.status.code().unwrap(),
                            std::str::from_utf8(&o.stdout).unwrap(),
                        )));
                    }

                    let _ = writeln!(&mut snapshot, "{}", last_output.unwrap_or_default());
                }

                fn stdout(root: &str, snapshot: &str) {
                    let root = PathBuf::from_str(root).unwrap();
                    let root = root.normalize();

                    let mut insta = insta::Settings::new();
                    insta.set_snapshot_path(root);
                    insta.set_prepend_module_to_snapshot(false);
                    insta.set_omit_expression(true);
                    let scope = insta.bind_to_scope();
                    insta::assert_snapshot!("stdout", snapshot);
                    drop(scope);
                }
                stdout(&format!("{}/{root}", repo_root.display()), &snapshot);

                Ok(())
            })
        })
        .collect();

    libtest_mimic::run(&args, tests).exit();
}

pub fn discover_tests(test_root: &Path) -> Vec<PathBuf> {
    use glob::glob;

    let mut entries = vec![];

    let pattern = format!("{}/**/snapshot.toml", test_root.display());
    for entry in glob(&pattern)
        .expect("Failed to read glob pattern")
        .flatten()
    {
        entries.push(entry.parent().unwrap().to_owned())
    }

    entries
}

fn clean_output(output: &str) -> String {
    #[derive(Default)]
    struct RawText(String);

    impl vte::Perform for RawText {
        fn print(&mut self, c: char) {
            self.0.push(c);
        }

        fn execute(&mut self, _: u8) {}

        fn hook(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn put(&mut self, b: u8) {
            self.0.push(b as char);
        }

        fn unhook(&mut self) {}

        fn osc_dispatch(&mut self, _: &[&[u8]], _: bool) {}

        fn csi_dispatch(&mut self, _: &vte::Params, _: &[u8], _: bool, _: char) {}

        fn esc_dispatch(&mut self, _: &[u8], _: bool, _: u8) {}
    }

    let mut raw = String::new();
    for line in output.lines() {
        let mut performer = RawText::default();
        let mut p = vte::Parser::new();
        for b in line.as_bytes() {
            p.advance(&mut performer, *b);
        }
        raw.push_str(&performer.0);
        raw.push('\n');
    }

    // Remove absolute paths from snapshot tests
    let manifest_dir = env!("CARGO_MANIFEST_DIR");
    let manifest_dir: PathBuf = PathBuf::from(manifest_dir);
    let parent = manifest_dir.parent().unwrap();
    let result = raw.replace(&format!("{}/", parent.display()), "");

    // Remove compilation time
    let r = Regex::new("(Finished (debug|release) \\[.*?\\] target\\(s\\) \\[.*?\\] in )(.*?s)")
        .unwrap();
    let result = r.replace(&result, "$1???");

    result.to_string()
}
