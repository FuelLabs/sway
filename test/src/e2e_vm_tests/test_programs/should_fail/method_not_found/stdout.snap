---
source: test/src/snapshot/mod.rs
assertion_line: 266
---
> forc build --path test/src/e2e_vm_tests/test_programs/should_fail/method_not_found
exit status: 1
output:
    Building test/src/e2e_vm_tests/test_programs/should_fail/method_not_found
   Compiling library std (test/src/e2e_vm_tests/reduced_std_libs/sway-lib-std-core)
   Compiling contract method_not_found (test/src/e2e_vm_tests/test_programs/should_fail/method_not_found)
warning
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:64:15
   |
62 | }
63 | 
64 | fn generic<T>(t: T) where T: TraitB {
   |               - This declaration is never used.
65 |   t.method();
66 |   t.method_b(42);
   |
____

warning
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:107:9
    |
105 |     s.method_b(42);
106 | 
107 |     let gs_u64 = GenericStruct::<u64> {};
    |         ------ This declaration is never used.
108 |     gs_u64.method_on_generic_struct();
109 | 
    |
____

warning
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:114:9
    |
112 |     generic(s);
113 | 
114 |     let caller = abi(Abi, 0x0000000000000000000000000000000000000000000000000000000000000000);
    |         ------ This declaration is never used.
115 |     caller.provided_fn_in_abi();
116 |     caller.provided_fn_in_abi(true);
    |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:65:5
   |
...
65 |   t.method();
   |     ^^^^^^ "method(T)" is not found for type "T".
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:66:5
   |
...
66 |   t.method_b(42);
   |     ^^^^^^^^ "method_b(T, {numeric})" is not found for type "T".
   |     -------- help: Only this "method_b" function is implemented for the type:
   |     -------- help: - T::method_b(self) -> ()
   |     -------- help: Did you mean to call that function?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:67:6
   |
...
67 |   T::associated();
   |      ^^^^^^^^^^ "associated()" is not found for type "T".
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:68:6
   |
...
68 |   T::associated_b(42);
   |      ^^^^^^^^^^^^ "associated_b({numeric})" is not found for type "T".
   |      ------------ help: Only this "associated_b" function is implemented for the type:
   |      ------------ help: - T::associated_b() -> ()
   |      ------------ help: Did you mean to call that function?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:91:31
   |
...
91 |     let _: ReturnedStruct = s.associated();
   |                               ^^^^^^^^^^ "associated(method_not_found::Struct) -> method_not_found::ReturnedStruct" is not found for type "method_not_found::Struct".
   |                               ---------- help: These "associated" functions are implemented for the type:
   |                               ---------- help: - <method_not_found::Struct as method_not_found::TraitA>::associated() -> method_not_found::ReturnedStruct
   |                               ---------- help: - method_not_found::Struct::associated() -> method_not_found::ReturnedStruct
   |                               ---------- help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:93:37
   |
...
93 |     let _: ReturnedStruct = Struct::method();
   |                                     ^^^^^^ "method() -> method_not_found::ReturnedStruct" is not found for type "method_not_found::Struct".
   |                                     ------ help: These "method" functions are implemented for the type:
   |                                     ------ help: - <method_not_found::Struct as method_not_found::TraitA>::method(self) -> ()
   |                                     ------ help: - method_not_found::Struct::method(self) -> ()
   |                                     ------ help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:95:30
   |
...
95 |     let _: ReturnedStruct =s.method(42);
   |                              ^^^^^^ "method(method_not_found::Struct, {numeric}) -> method_not_found::ReturnedStruct" is not found for type "method_not_found::Struct".
   |                              ------ help: These "method" functions are implemented for the type:
   |                              ------ help: - <method_not_found::Struct as method_not_found::TraitA>::method(self) -> ()
   |                              ------ help: - method_not_found::Struct::method(self) -> ()
   |                              ------ help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:96:36
   |
...
96 |     let _: ReturnedStruct =Struct::associated(42);
   |                                    ^^^^^^^^^^ "associated({numeric}) -> method_not_found::ReturnedStruct" is not found for type "method_not_found::Struct".
   |                                    ---------- help: These "associated" functions are implemented for the type:
   |                                    ---------- help: - <method_not_found::Struct as method_not_found::TraitA>::associated() -> method_not_found::ReturnedStruct
   |                                    ---------- help: - method_not_found::Struct::associated() -> method_not_found::ReturnedStruct
   |                                    ---------- help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:98:7
   |
...
98 |     s.different_method(true);
   |       ^^^^^^^^^^^^^^^^ "different_method(method_not_found::Struct, bool)" is not found for type "method_not_found::Struct".
   |       ---------------- help: These "different_method" functions are implemented for the type:
   |       ---------------- help: - <method_not_found::Struct as method_not_found::TraitA>::different_method(self) -> ()
   |       ---------------- help: - method_not_found::Struct::different_method(self, _x: u64) -> ()
   |       ---------------- help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
  --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:99:13
   |
...
99 |     Struct::different_associated(true);
   |             ^^^^^^^^^^^^^^^^^^^^ "different_associated(bool)" is not found for type "method_not_found::Struct".
   |             -------------------- help: These "different_associated" functions are implemented for the type:
   |             -------------------- help: - <method_not_found::Struct as method_not_found::TraitA>::different_associated() -> ()
   |             -------------------- help: - method_not_found::Struct::different_associated(_x: u64) -> ()
   |             -------------------- help: Did you mean to call one of those functions?
   |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:101:13
    |
...
101 |     Struct::associated_with_params();
    |             ^^^^^^^^^^^^^^^^^^^^^^ "associated_with_params()" is not found for type "method_not_found::Struct".
    |             ---------------------- help: These "associated_with_params" functions are implemented for the type:
    |             ---------------------- help: - <method_not_found::Struct as method_not_found::TraitA>::associated_with_params(_x: u64) -> ()
    |             ---------------------- help: - method_not_found::Struct::associated_with_params(_x: u64) -> ()
    |             ---------------------- help: Did you mean to call one of those functions?
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:103:7
    |
...
103 |     s.method_with_params();
    |       ^^^^^^^^^^^^^^^^^^ "method_with_params(method_not_found::Struct)" is not found for type "method_not_found::Struct".
    |       ------------------ help: These "method_with_params" functions are implemented for the type:
    |       ------------------ help: - <method_not_found::Struct as method_not_found::TraitA>::method_with_params(self, _x: u64) -> ()
    |       ------------------ help: - method_not_found::Struct::method_with_params(self, _x: u64) -> ()
    |       ------------------ help: Did you mean to call one of those functions?
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:105:7
    |
...
105 |     s.method_b(42);
    |       ^^^^^^^^ "method_b(method_not_found::Struct, {numeric})" is not found for type "method_not_found::Struct".
    |       -------- help: These "method_b" functions are implemented for the type:
    |       -------- help: - <method_not_found::Struct as method_not_found::TraitB>::method_b(self) -> ()
    |       -------- help: - <method_not_found::Struct as method_not_found::TraitBB>::method_b(self, _x: u64, _y: u64) -> ()
    |       -------- help: Did you mean to call one of those functions?
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:108:12
    |
...
108 |     gs_u64.method_on_generic_struct();
    |            ^^^^^^^^^^^^^^^^^^^^^^^^ "method_on_generic_struct(method_not_found::GenericStruct<u64>)" is not found for type "method_not_found::GenericStruct<u64>".
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:110:27
    |
...
110 |     GenericStruct::<u64>::associated_on_generic_struct();
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ "associated_on_generic_struct()" is not found for type "method_not_found::GenericStruct<u64>".
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:115:12
    |
...
115 |     caller.provided_fn_in_abi();
    |            ^^^^^^^^^^^^^^^^^^ "provided_fn_in_abi(ContractCaller<Abi>)" is not found for type "ContractCaller<Abi>".
    |            ------------------ help: Only this "provided_fn_in_abi" function is implemented for the type:
    |            ------------------ help: - Abi::provided_fn_in_abi(x: u64) -> ()
    |            ------------------ help: Did you mean to call that function?
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:116:12
    |
...
116 |     caller.provided_fn_in_abi(true);
    |            ^^^^^^^^^^^^^^^^^^ "provided_fn_in_abi(ContractCaller<Abi>, bool)" is not found for type "ContractCaller<Abi>".
    |            ------------------ help: Only this "provided_fn_in_abi" function is implemented for the type:
    |            ------------------ help: - Abi::provided_fn_in_abi(x: u64) -> ()
    |            ------------------ help: Did you mean to call that function?
    |
____

error: Associated function or method is not found
   --> test/src/e2e_vm_tests/test_programs/should_fail/method_not_found/src/main.sw:118:17
    |
...
118 |     let _ = 0u8 + true;
    |                 ^ "add(u8, bool)" is not found for type "u8".
    |                 - help: Only this "add" function is implemented for the type:
    |                 - help: - <u8 as std::ops::Add>::add(self, other: u8) -> u8
    |                 - help: Did you mean to call that function?
    |
____

  Aborting due to 18 errors.
error: Failed to compile method_not_found
